### PR TITLE
feat(siwe): expose a SIWE client

### PR DESCRIPTION
**Pull request by satish-ravi**
- Made a new SIWE client and pulled in auth and clock specific methods and included a generic method for fetch after login
- Updated fiat connect client to use SIWE client

### CODE REVIEW

File: src/fiat-connect-client.ts [hunk 1]
```diff
@@ -190,71 +122,39 @@ export class FiatConnectClientImpl implements FiatConnectApiClient {
...
    * Returns an approximation of the current server time, taking into account clock differences
    * between client and server. Returns the earliest possible server time based on the max error
    * of the clock diff between client and server, to ensure that sessions created using this time
    * are not issued in the future with respect to the server clock.
    */
   async getServerTimeApprox(): Promise<Result<Date, ResponseError>> {
-    const clockDiffResponse = await this.getClockDiffApprox()
-    if (!clockDiffResponse.isOk) {
-      return Result.err(clockDiffResponse.error)
+    try {
```
Comment (cajubelt) [inline src/fiat-connect-client.ts:132]: within this codebase, this is only used by `login`, which has logic that is more specific to SIWE clients than FiatConnect clients. How about moving the getServerTimeApprox logic to the SIWE client too?

Comment (cajubelt) [reply src/fiat-connect-client.ts:132]: (before taking my suggestion, it's probably worth looking at consumers of the SDK and making sure they don't use it)

**Review by awtsui — COMMENTED** [review]
Left a question but besides LGTM

File: src/fiat-connect-client.ts [hunk 2]
```diff
@@ -190,71 +122,39 @@ export class FiatConnectClientImpl implements FiatConnectApiClient {
...
     try {
-      const response = await this.fetchImpl(`${this.config.baseUrl}/clock`, {
-        method: 'GET',
-        headers: this._getAuthHeader(),
-      })
-      const data = await response.json()
-      if (!response.ok) {
-        return handleError(data)
-      }
+      const data = await this._siweClient.getClock()
```
Comment (cajubelt) [inline src/fiat-connect-client.ts:157]: same here, seems like we can probly remove getClock from the FC client

File: src/types.ts [hunk 3]
```diff
@@ -59,10 +69,24 @@ export interface FiatConnectApiClient {
...
+  loginUrl: string
+  clockUrl: string
+}
+
 export interface LoginParams {
   issuedAt?: Date
 }
 
+export type SiweLoginParams = LoginParams & {
+  headers?: Record<string, string>
```
Comment (awtsui) [inline src/types.ts:87]: What kind of header do we expect to add?

Comment (satish-ravi) [reply src/types.ts:87]: The auth header with the client api key: https://github.com/fiatconnect/fiatconnect-sdk/blob/main/src/fiat-connect-client.ts#L62-L66

File: src/siwe-client.ts [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    this._sessionExpiry = expirationTime
+  }
+
+  /**
+   * Checks if a logged in session exists.
+   *
+   * @returns true if an unexpired session exists, else false
+   */
+  isLoggedIn(): boolean {
+    return !!(this._sessionExpiry && this._sessionExpiry > new Date())
```
Comment (cajubelt) [inline src/siwe-client.ts:100]: seems like it might be worth adding a check for a cookie existing, too

**Review by cajubelt — COMMENTED** [review]
good idea!

commented in a couple places where we might be able to drop SIWE-specific logic from the FC client class

File: src/fiat-connect-client.ts (hunk 1 above)
Comment (satish-ravi) [reply src/fiat-connect-client.ts:132]: the clock related methods in fc-client are just wrappers around the siwe client methods. I just kept them for backwards compatibility. The wallet (which I believe is the only consumer) doesn't use any of the clock methods directly. So I even considered removing these from the FiatConnectApiClient / SiweApiClient interfaces. Any other consumers we need to check? cc @jophish

File: src/siwe-client.ts (hunk 4 above)
Comment (satish-ravi) [reply src/siwe-client.ts:100]: Good idea, I think we can even ensure the cookie hasn't expired over the assumed sessionExpiry value. But the current cookie handling is broken, @awtsui is working on fixing it on top of this PR. We can do this in a follow up

File: src/fiat-connect-client.ts (hunk 2 above)
Comment (jophish) [reply src/fiat-connect-client.ts:157]: I think it's good to keep `getClock` on the FC client; without it, there's no interface for clients to hit `/clock`, which is a required FC endpoint, so seems good to keep support for it in the FC SDK.

File: src/fiat-connect-client.ts [hunk 5]
```diff
@@ -106,58 +88,8 @@ export class FiatConnectClientImpl implements FiatConnectApiClient {
...
-        nonce: generateNonce(),
-        issuedAt: issuedAt.toISOString(),
-        expirationTime: expirationTime.toISOString(),
-      })
-      const message = siweMessage.prepareMessage()
-      const body: AuthRequestBody = {
-        message,
-        signature: await this.signingFunction(message),
-      }
+      await this._siweClient.login({ issuedAt, headers: this._getAuthHeader() })
```
Comment (jophish) [inline src/fiat-connect-client.ts:91]: Can we move the clock coordination/sync stuff above into the SIWE client?

File: src/fiat-connect-client.ts (hunk 1 above)
Comment (jophish) [reply src/fiat-connect-client.ts:132]: We currently don't use `getServerTimeApprox` or `getClockDiffApprox` outside of the FC SDK, but it's possible that a client may want to use them in order to present accurate expirations for quotes to an end user. For that reason, it's probably not a bad idea to continue to expose them via the FiatConnectApiClient interface as a convenience, and in turn, the `SiweApiClient` interface.

File: src/fiat-connect-client.ts (hunk 5 above)
Comment (satish-ravi) [reply src/fiat-connect-client.ts]: oops completely missed this, should be fixed now.

File: src/fiat-connect-client.ts (hunk 1 above)
Comment (cajubelt) [reply src/fiat-connect-client.ts:122]: fiatconnect validate is another consumer to check

Comment (cajubelt) [reply src/fiat-connect-client.ts:122]: up to you whether you want to remove it or not

**Review by cajubelt — APPROVED** [review]

**Pull request merged**